README.md
![License](https://img.shields.io/github/license/mygamedevtools/fps-counter)
[![openupm](https://img.shields.io/npm/v/com.mygamedevtools.fps-counter?label=openupm&registry_uri=https://package.openupm.com)](https://openupm.com/packages/com.mygamedevtools.fps-counter/)
![Release](https://img.shields.io/github/v/release/mygamedevtools/fps-counter?sort=semver)
![Last Commit](https://img.shields.io/github/last-commit/mygamedevtools/fps-counter)

FPS Counter
===

_A package to measure the average, highest and lowest frames per second in any Unity Application._

Installation
---

### OpenUPM

This package is available on the [OpenUPM](https://openupm.com/packages/com.mygamedevtools.fps-counter) registry. Add the package via the [openupm-cli](https://github.com/openupm/openupm-cli):

```
openupm add com.mygamedevtools.fps-counter
```

### [Installing from Git](https://docs.unity3d.com/Manual/upm-ui-giturl.html) _(requires [Git](https://git-scm.com/) installed and added to the PATH)_

1. Open `Edit/Project Settings/Package Manager`.
2. Click <kbd>+</kbd>.
3. Select `Add package from git URL...`.
4. Paste `com.mygamedevtools.fps-counter` into name.
5. Click `Add`.

Usage
---

Add the component `FPSCounter` to any `GameObject` by manually dragging the script or via `Add Component/Stats/FPS Counter`. 
Then, add a `FPSDisplay` component to the same `GameObject` the same way you just did with `FPSCounter`.
Now, you'll need to assign **3** `TextMeshProUGUI` objects in the `FPSDisplay` inspector to serve as output for the **average**, **highest** and **lowest** FPS counts respectively.

You can also configure the other settings to fit your needs:
- The `Color Gradient` can alter the label color based on the target FPS where the **leftmost** value is the **closest** FPS value to **0**, 
and the **rightmost** value is the **closest** FPS value to the `Target Framerate`.
- The `Refresh Rate` controls how often the `FPSDisplay` updates its values.
- The `Target Framerate` is a guide to the `Color Gradient` property. It will be the reference for the value of 1 in the color gradient.
- The `Max Tracked Fps` controls what should be the maximum FPS to track.

Advanced Usage
---

The core FPS counter feature lies in the `FPSBuffer`. 
It needs a `bufferSize` to be provided on creation, and then it will fill this buffer with FPS values to calculate the average, highest and lowest framerates.
You need to call `UpdateBuffer` every Unity `Update` to accurately calculate the FPS.
The `FPSCounter` component simply creates a `FPSBuffer` object on `Awake` with the provided `Buffer Size` in the inspector, 
and runs `FPSBuffer.UpdateBuffer` on its `Update`. 
You can use any other structure as an alternative to the `FPSCounter` or even integrate to your existing systems just by using a `FPSBuffer` object.

The `FPSDisplay` component reads from the `FPSCounter` component directly and outputs the average, highest and lowest fps.
You can also create another component to display the values either from the `FPSCounter` or from another system you created that uses a `FPSBuffer` object.
The `FPSDisplay` class was created with an array of strings with numbers up to **300**, to avoid string allocation in runtime.

---

Don't hesitate to create [issues](https://github.com/mygamedevtools/fps-counter/issues) for suggestions and bugs. Have fun!


CHANGELOG.md
# [1.5.0](https://github.com/mygamedevtools/fps-counter/compare/1.4.1...1.5.0) (2022-10-25)


### Features

* Add CI configuration ([cbaa4b0](https://github.com/mygamedevtools/fps-counter/commit/cbaa4b08757eb0b8ee64c464653086ea14fcc73d))

## [1.4.1] - 2022-10-24
- Added: OpenUPM documentation.

## [1.4.0] - 2022-10-03
- Changed: Updated organization name to comply with [Unity Package Guidelines](https://unity.com/legal/terms-of-service/software/package-guidelines).
- Fixed: Updated assembly definition and namespaces names to reflect the organization name changes.

## [1.3.0] - 2022-09-13
- Changed: Moved repository to My Unity Tools organization.
- Changed: Updated package name and author.

## [1.2.1] - 2022-09-09
- Added: Support for [hot reloading](https://docs.unity3d.com/2019.3/Documentation/Manual/script-Serialization.html) in the Unity Editor.

## [1.2.0] - 2021-02-22
- Changed: Refactored scripts to the My Unity Tools namespace.

## [1.1.0] - 2021-02-05
- Changed: Improved `FPSDisplay` component flexibility.

## [1.0.0] - 2021-01-31
- Initial release.

[1.4.1]: https://github.com/mygamedevtools/fps-counter/compare/1.4.0...1.4.1
[1.4.0]: https://github.com/mygamedevtools/fps-counter/compare/1.3.0...1.4.0
[1.3.0]: https://github.com/mygamedevtools/fps-counter/compare/1.2.1...1.3.0
[1.2.1]: https://github.com/mygamedevtools/fps-counter/compare/1.2.0...1.2.1
[1.2.0]: https://github.com/mygamedevtools/fps-counter/compare/1.1.0...1.2.0
[1.1.0]: https://github.com/mygamedevtools/fps-counter/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/mygamedevtools/fps-counter/compare/330cde1...1.0.0


Runtime/UI/FPSDisplay.cs
/**
 * FPSDisplay.cs
 * Created by: João Borks [joao.borks@gmail.com]
 * Created on: 1/30/2021 (en-US)
 */

using System.Collections;
using UnityEngine;
using UnityEngine.Serialization;

namespace MyGameDevTools.Stats.UI
{
    [RequireComponent(typeof(FPSCounter))]
    [AddComponentMenu("Stats/FPS Display")]
    public class FPSDisplay : MonoBehaviour
    {
        /// <summary>
        /// Collection of all supported number strings. It's predefined to avoid allocation in runtime
        /// </summary>
        readonly string[] _numberStrings =
        {
            "0", "1", "2", "3", "4", "5", "6", "7", "8", "9",
            "10", "11", "12", "13", "14", "15", "16", "17", "18", "19",
            "20", "21", "22", "23", "24", "25", "26", "27", "28", "29",
            "30", "31", "32", "33", "34", "35", "36", "37", "38", "39",
            "40", "41", "42", "43", "44", "45", "46", "47", "48", "49",
            "50", "51", "52", "53", "54", "55", "56", "57", "58", "59",
            "60", "61", "62", "63", "64", "65", "66", "67", "68", "69",
            "70", "71", "72", "73", "74", "75", "76", "77", "78", "79",
            "80", "81", "82", "83", "84", "85", "86", "87", "88", "89",
            "90", "91", "92", "93", "94", "95", "96", "97", "98", "99",
            "100", "101", "102", "103", "104", "105", "106", "107", "108", "109",
            "110", "111", "112", "113", "114", "115", "116", "117", "118", "119",
            "120", "121", "122", "123", "124", "125", "126", "127", "128", "129",
            "130", "131", "132", "133", "134", "135", "136", "137", "138", "139",
            "140", "141", "142", "143", "144", "145", "146", "147", "148", "149",
            "150", "151", "152", "153", "154", "155", "156", "157", "158", "159",
            "160", "161", "162", "163", "164", "165", "166", "167", "168", "169",
            "170", "171", "172", "173", "174", "175", "176", "177", "178", "179",
            "180", "181", "182", "183", "184", "185", "186", "187", "188", "189",
            "190", "191", "192", "193", "194", "195", "196", "197", "198", "199",
            "200", "201", "202", "203", "204", "205", "206", "207", "208", "209",
            "210", "211", "212", "213", "214", "215", "216", "217", "218", "219",
            "220", "221", "222", "223", "224", "225", "226", "227", "228", "229",
            "230", "231", "232", "233", "234", "235", "236", "237", "238", "239",
            "240", "241", "242", "243", "244", "245", "246", "247", "248", "249",
            "250", "251", "252", "253", "254", "255", "256", "257", "258", "259",
            "260", "261", "262", "263", "264", "265", "266", "267", "268", "269",
            "270", "271", "272", "273", "274", "275", "276", "277", "278", "279",
            "280", "281", "282", "283", "284", "285", "286", "287", "288", "289",
            "290", "291", "292", "293", "294", "295", "296", "297", "298", "299",
            "+300"
        };

        [Header("References")]
        [SerializeField, FormerlySerializedAs("averageLabel")]
        FPSLabel _averageLabel;
        [SerializeField, FormerlySerializedAs("highestLabel")]
        FPSLabel _highestLabel;
        [SerializeField, FormerlySerializedAs("lowestLabel")]
        FPSLabel _lowestLabel;

        [Header("Settings")]
        [SerializeField, Range(.05f, 3), FormerlySerializedAs("refreshRate")]
        float _refreshRate = .2f;
        [SerializeField, Range(30, 300), FormerlySerializedAs("targetFramerate")]
        int _targetFramerate = 60;
        [SerializeField, Range(30, 300), FormerlySerializedAs("maxTrackedFPS")]
        int _maxTrackedFPS = 300;

        FPSLabel[] _labels;
        FPSCounter _counter;

        void OnEnable()
        {
            _counter = GetComponent<FPSCounter>();
            _labels = new FPSLabel[]
            {
                _averageLabel,
                _highestLabel,
                _lowestLabel
            };
            UpdateDisplay();
        }

        Coroutine UpdateDisplay()
        {
            return StartCoroutine(updateDisplayRoutine());
            IEnumerator updateDisplayRoutine()
            {
                var delay = new WaitForSeconds(_refreshRate);
                yield return delay;
                int i;
                int length = _labels.Length;
                int fps;
                while (true)
                {
                    for (i = 0; i < length; i++)
                    {
                        fps = _counter.FPSBuffer.Values[i];
                        _labels[i].UpdateLabel(_numberStrings[Mathf.Clamp(fps, 0, _maxTrackedFPS)], (float)fps / _targetFramerate);
                    }
                    yield return delay;
                }
            }
        }
    }
}